### PR TITLE
CHNS Free Energy Postprocessing

### PR DESCRIPTION
**Pull request by PierreLaurentinCS**
# Description of the problem

The Cahn-Hilliard equations are energy based equations for solving multiphase flows : the interface is the result of two energies in competition, bulk and interface energy. The system, which starts from a non-equilibrium state, lowers its total energy (bulk + interface) to reach equilibrium. In order to visualize this process I added the energy computation into the postprocessing section.
Very minor : added the computation of the volume of each phases.

# Description of the solution

- New functions `calculate_phase_energy()` and `write_phase_energy` were implemented as well as new parameter entries in the postprocess subsection : `compute phase energy` (boolean) and `phase energy name`

# How Has This Been Tested?

- A new test was added : `cahn_hilliard_navier_stokes_energy_postprocessing`

# Documentation

- Postprocessing documentation was updated to include this new feature

- File modified : `post_processing.rst`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,12 @@
 All notable changes to the Lethe project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/).
 
+## [Master] - 2024-04-17
+
+### Added
+
+- MINOR Added computation of the energies involved in the Cahn-Hilliard equations and updated related documentation. [#1095](https://github.com/lethe-cfd/lethe/pull/1095)
```
Comment (voferreira) [inline CHANGELOG.md:10]:
```suggestion
- MINOR Added Cahn-Hilliard equations energy computation and output. The documentation was updated accordingly. [#1095](https://github.com/lethe-cfd/lethe/pull/1095)
```

File: applications_tests/lethe-fluid/cahn_hilliard_navier_stokes_energy_postprocessing.prm [hunk 2]
```diff
@@ -0,0 +1,127 @@
+set dimension = 2
+
```
Comment (voferreira) [inline applications_tests/lethe-fluid/cahn_hilliard_navier_stokes_energy_postprocessing.prm:2]: We don't add the subsection comments anymore? (Like, the ***************)

File: doc/source/parameters/cfd/post_processing.rst [hunk 3]
```diff
@@ -67,6 +67,8 @@ This subsection controls the post-processing other than the forces and torque on
     # Other Cahn-Hilliard postprocessing
     set calculate phase statistics       = false
     set phase statistics name            = phase_statistics
+    set calculate phase energy           = true
```
Comment (voferreira) [inline doc/source/parameters/cfd/post_processing.rst:70]: True by default? It should be false, no?

Comment (voferreira) [reply doc/source/parameters/cfd/post_processing.rst:70]: What do you think about removing the word "calculate" from the parameters? (can be another pr)

File: doc/source/parameters/cfd/post_processing.rst [hunk 4]
```diff
@@ -219,8 +221,28 @@ This subsection controls the post-processing other than the forces and torque on
 
 * ``mass conservation name``: name of the output file containing the mass of both fluids for VOF simulations. The default file name is ``mass_conservation_information``.
   
-* ``calculate phase statistics``: outputs phase statistics from the solution of the Cahn-Hilliard equations, including minimum, maximum, average, and standard deviation of the phase order parameter. This works only with the :doc:`cahn_hilliard` solver.
+* ``calculate phase statistics``: outputs phase statistics from the solution of the Cahn-Hilliard equations, including minimum, maximum, average, integral of the phase order parameter. Also includes the volume of each phase. This works only with the :doc:`cahn_hilliard` solver.
```
Comment (voferreira) [inline doc/source/parameters/cfd/post_processing.rst:224]:
```suggestion
* ``calculate phase statistics``: outputs phase statistics from the solution of the Cahn-Hilliard equations, including minimum, maximum, average, and integral of the phase order parameter. The volume of each phase is also computed. This works only with the :doc:`cahn_hilliard` solver.
```

Comment (voferreira) [reply doc/source/parameters/cfd/post_processing.rst:224]: " This works only with the :doc:`cahn_hilliard` solver." can be a warning

Comment (voferreira) [reply doc/source/parameters/cfd/post_processing.rst:224]: I would rephrase it to: outputs Cahn-Hilliard phase statistics, including .............

File: doc/source/parameters/cfd/post_processing.rst [hunk 5]
```diff
@@ -219,8 +221,28 @@ This subsection controls the post-processing other than the forces and torque on
...
 
 * ``phase statistics name``: name of the output file containing phase order parameter statistics from Cahn-Hilliard simulations. The default file name is ``phase_statistics``.
 
+* ``calculate phase energy``:  outputs phase energies from the solution of the Cahn-Hilliard equations : bulk energy, interface energy and total energy. This works only with the :doc:`cahn_hilliard` solver. The energies are computed as follow:
```
Comment (voferreira) [inline doc/source/parameters/cfd/post_processing.rst:228]:
```suggestion
* ``calculate phase energy``:  outputs phase energies from the solution of the Cahn-Hilliard equations, including: bulk energy, interface energy, and total energy. This works only with the :doc:`cahn_hilliard` solver. The energies are computed as:
```

Comment (voferreira) [reply doc/source/parameters/cfd/post_processing.rst:228]: Same comment about the warning

File: doc/source/parameters/cfd/post_processing.rst [hunk 6]
```diff
@@ -219,8 +221,28 @@ This subsection controls the post-processing other than the forces and torque on
...
+      
+  .. math::
+
+     E_{interface} = \int_{\Omega} 0.5\epsilon^2|\nabla \phi |^2 \mathrm{d}\Omega 
+      
+  .. math::
+
+     E_{total} = E_{bulk} + E_{interface}  
+    
+  where :math:`\epsilon` is the numerical interface thickness. Note that these energies are not homogeneous to physical energies, but they give nonetheless a good insight on the system's evolution.
```
Comment (voferreira) [inline doc/source/parameters/cfd/post_processing.rst:242]:
```suggestion
  where :math:`\epsilon` is the numerical interface thickness. Note that these energies are not homogeneous to physical energies. Nonetheless, they are a convenient way to track the system's evolution.
```

File: doc/source/parameters/cfd/post_processing.rst [hunk 7]
```diff
@@ -219,8 +221,28 @@ This subsection controls the post-processing other than the forces and torque on
...
+
+
+* ``phase energy name``: name of the output file containing phase energies from Cahn-Hilliard simulations. The default file name is ``phase_energy``.
```
Comment (voferreira) [inline doc/source/parameters/cfd/post_processing.rst:245]: Is there an extension to this file?
Where is it stored? at the output folder?
Could you add this information here?

File: include/solvers/cahn_hilliard.h [hunk 8]
```diff
@@ -402,11 +402,24 @@ class CahnHilliard : public AuxiliaryPhysics<dim, GlobalVectorType>
...
   /**
-   * @brief Writes the phase integral to an output file
+   * @brief Writes the phase statistics to an output file
    */
   void
   write_phase_statistics();
 
+  /**
+   *
+   * @return A pointer to an array of double containing the energies. First element is bulk energy, second is interface energy and third is total energy
```
Comment (voferreira) [inline include/solvers/cahn_hilliard.h:412]: This is not a return function. It is void. Does it make sense to use return here?

File: source/core/parameters.cc [hunk 9]
```diff
@@ -1828,6 +1828,18 @@ namespace Parameters
         "mass_conservation_information",
         Patterns::FileName(),
         "Name of mass conservation output file in VOF simulations");
+
+      prm.declare_entry(
+        "calculate phase energy",
+        "false",
```
Comment (voferreira) [inline source/core/parameters.cc:1834]: So it is false by default. Can you please change it in the documentation?

File: source/core/parameters.cc [hunk 10]
```diff
@@ -1828,6 +1828,18 @@ namespace Parameters
...
+        Patterns::Bool(),
+        "Enable calculation of phase energies : total energy, bulk energy and interface energy");
```
Comment (voferreira) [inline source/core/parameters.cc:1836]:
```suggestion
        "Enable calculation of phase energies, including: total energy, bulk energy, and interface energy");
```

File: source/core/parameters.cc [hunk 11]
```diff
@@ -1828,6 +1828,18 @@ namespace Parameters
...
+
+      prm.declare_entry(
+        "phase energy name",
+        "phase_energy",
+        Patterns::FileName(),
+        "Name of energy output file in Cahn-Hilliard simulations.");
```
Comment (voferreira) [inline source/core/parameters.cc:1842]: Comment on the output path

**Review by voferreira — APPROVED** [review]
Good job :)
Very few and tiny suggestions but it looks good to me

File: applications_tests/lethe-fluid/cahn_hilliard_navier_stokes_energy_postprocessing.prm (hunk 2 above)
Comment (PierreLaurentinCS) [reply applications_tests/lethe-fluid/cahn_hilliard_navier_stokes_energy_postprocessing.prm:2]: Agreed but some tests have it and some other don't, I would say that the test's .prm must be as minimalists as possible, hence the absence of subsection decorations. Can be discussed though

File: doc/source/parameters/cfd/post_processing.rst (hunk 3 above)
Comment (PierreLaurentinCS) [reply doc/source/parameters/cfd/post_processing.rst:70]: should definitely be false, you're right. For the calculate I think I'll leave it to another PR

**Review by blaisb — APPROVED** [review]
Nothing to add here, good work :).  I'm a bit alarmed by the growing number of post-processing parameters, but I presume this is normal as we increase the number of physics.
